base/lock.jl

```diff
@@ -201,18 +294,38 @@ internal counter and return immediately.
         rl.reentrancy_cnt = n
         if n == 0x0000_00000
             @atomic :monotonic rl.locked_by = nothing
-            if (@atomicswap :release rl.havelock = 0x00) == 0x02
+            result = (@atomicreplace :release :monotonic rl.havelock LOCKED_BIT => 0x00)
+            if result.success
+                return true
+            else
                 (@noinline function notifywaiters(rl)
                     cond_wait = rl.cond_wait
                     lock(cond_wait)
                     try
-                        notify(cond_wait)
+
+                        tasks_notified = notify(cond_wait, all=false)
+                        if tasks_notified == 0
+                            # Either:
+                            # - We won the race to the `cond_wait` lock as a task was about to park
+                            # - We are the last task on the queue
+                            #
+                            # Unlock anyway as any parking task will retry
+
+                            @atomic :release rl.havelock = 0x00
+                        elseif isempty(cond_wait.waitq)
+                            # We notified the last task, unlock and unset the parked bit
+                            @atomic :release rl.havelock = 0x00
+                        else
+                            # There are more tasks on the queue, we unlock and keep the parked bit set
+                            @atomic :release rl.havelock = PARKED_BIT
+                            notify(cond_wait, all=false)
```

Comment:
There's already a `notify` call above; isn't this one going to unnecessarily schedule another task?

Reply:
Good catch! It is leftover from previous refactoring, the new proposal removes that.